In [1]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
gdp_df = pd.read_csv('../data/gdp_percapita.csv')

In [3]:
gdp_df.head()

Country or Area  Year        Value  Value Footnotes
0     Afghanistan  2020  1970.560169              NaN
1     Afghanistan  2019  2065.036235              NaN
2     Afghanistan  2018  2033.804389              NaN
3     Afghanistan  2017  2058.400221              NaN
4     Afghanistan  2016  2057.067978              NaN

In [4]:
gdp_df.tail()

Country or Area  Year        Value  Value Footnotes
7171        Zimbabwe  1994  2603.927369              NaN
7172        Zimbabwe  1993  2420.092791              NaN
7173        Zimbabwe  1992  2437.152908              NaN
7174        Zimbabwe  1991  2733.696626              NaN
7175        Zimbabwe  1990  2652.129209              NaN

# #5. How many rows and columns are in `gdp_df`? What are the data types of each column?

In [5]:
gdp_df.shape

(7176, 4)

In [6]:
gdp_df.columns

Index(['Country or Area', 'Year', 'Value', 'Value Footnotes'], dtype='object')

  # There are 7176 rows and 4 columns. The columns are all data type object.

# #6. Drop the `Value Footnotes` column and rename the remaining three to 'Country', 'Year', and 'GDP_Per_Capita'.

In [7]:
#gdp_df2 = gdp_df.columns.drop('Value Footnotes')

In [8]:
gdp_df2 = gdp_df.rename(columns={'Country or Area':'Country', 'Year':'Year', 'Value':'GDP_Per_Capita'})
gdp_df2

Country  Year  GDP_Per_Capita  Value Footnotes
0     Afghanistan  2020     1970.560169              NaN
1     Afghanistan  2019     2065.036235              NaN
2     Afghanistan  2018     2033.804389              NaN
3     Afghanistan  2017     2058.400221              NaN
4     Afghanistan  2016     2057.067978              NaN
...           ...   ...             ...              ...
7171     Zimbabwe  1994     2603.927369              NaN
7172     Zimbabwe  1993     2420.092791              NaN
7173     Zimbabwe  1992     2437.152908              NaN
7174     Zimbabwe  1991     2733.696626              NaN
7175     Zimbabwe  1990     2652.129209              NaN

[7176 rows x 4 columns]

In [9]:
gdp_df3 = gdp_df2.drop(columns = 'Value Footnotes')

In [10]:
gdp_df3

Country  Year  GDP_Per_Capita
0     Afghanistan  2020     1970.560169
1     Afghanistan  2019     2065.036235
2     Afghanistan  2018     2033.804389
3     Afghanistan  2017     2058.400221
4     Afghanistan  2016     2057.067978
...           ...   ...             ...
7171     Zimbabwe  1994     2603.927369
7172     Zimbabwe  1993     2420.092791
7173     Zimbabwe  1992     2437.152908
7174     Zimbabwe  1991     2733.696626
7175     Zimbabwe  1990     2652.129209

[7176 rows x 3 columns]

# #7. How many countries have data for all years? Which countries are missing many years of data? Look at the number of observations per year. What do you notice? 

In [11]:
year_count = gdp_df3.Year.value_counts()
year_count

2018    242
2017    242
2016    242
2015    242
2014    242
2013    242
2019    241
2020    240
2012    240
2011    240
2010    239
2009    239
2008    238
2007    237
2006    237
2004    236
2005    236
2003    235
2002    235
2001    234
2000    233
1999    227
1998    226
1997    226
1996    222
1995    222
1994    211
1993    209
1992    208
1991    207
1990    206
Name: Year, dtype: int64

In [12]:
year_count.head()

2018    242
2017    242
2016    242
2015    242
2014    242
Name: Year, dtype: int64

In [13]:
type(year_count)

pandas.core.series.Series

In [14]:
year_count = year_count.to_frame()

In [15]:
type(year_count)

pandas.core.frame.DataFrame

In [16]:
year_count


Year
2018   242
2017   242
2016   242
2015   242
2014   242
2013   242
2019   241
2020   240
2012   240
2011   240
2010   239
2009   239
2008   238
2007   237
2006   237
2004   236
2005   236
2003   235
2002   235
2001   234
2000   233
1999   227
1998   226
1997   226
1996   222
1995   222
1994   211
1993   209
1992   208
1991   207
1990   206

In [27]:
gdp_df3.sort_values(by = 'Year').Country

7175               Zimbabwe
5688             Seychelles
4279             Micronesia
1065               Cameroon
3688                Lebanon
               ...         
2395                 Greece
3851    Lower middle income
821                  Brunei
6065              St. Lucia
0               Afghanistan
Name: Country, Length: 7176, dtype: object

In [28]:
gdp_df3.sort_values(['Country','Year'])

Country  Year  GDP_Per_Capita
18    Afghanistan  2002     1189.784668
17    Afghanistan  2003     1235.810063
16    Afghanistan  2004     1200.278013
15    Afghanistan  2005     1286.793659
14    Afghanistan  2006     1315.789117
...           ...   ...             ...
7149     Zimbabwe  2016     2341.304920
7148     Zimbabwe  2017     2416.049969
7147     Zimbabwe  2018     2497.136618
7146     Zimbabwe  2019     2310.634802
7145     Zimbabwe  2020     2134.555461

[7176 rows x 3 columns]

In [30]:
year_count2=gdp_df3.value_counts(['Country','Year'])
year_count2

Country       Year
Afghanistan   2002    1
OECD members  2018    1
Nigeria       2000    1
              1999    1
              1998    1
                     ..
Ghana         2014    1
              2013    1
              2012    1
              2011    1
Zimbabwe      2020    1
Length: 7176, dtype: int64

In [31]:
type(year_count2)

pandas.core.series.Series

In [32]:
year_count2=year_count2.to_frame()

In [33]:
type(year_count2)

pandas.core.frame.DataFrame

In [34]:
year_count2

0
Country      Year   
Afghanistan  2002  1
OECD members 2018  1
Nigeria      2000  1
             1999  1
             1998  1
...               ..
Ghana        2014  1
             2013  1
             2012  1
             2011  1
Zimbabwe     2020  1

[7176 rows x 1 columns]

In [43]:
year_count3 =pd.merge(year_count, gdp_df3, how='outer', on=['Year','Year'])

In [44]:
year_count3

Year   Country  GDP_Per_Capita
0      242       NaN             NaN
1      242       NaN             NaN
2      242       NaN             NaN
3      242       NaN             NaN
4      242       NaN             NaN
...    ...       ...             ...
7202  1990   Vanuatu     2855.554643
7203  1990   Vietnam     2066.372215
7204  1990     World     9704.621768
7205  1990    Zambia     2190.181724
7206  1990  Zimbabwe     2652.129209

[7207 rows x 3 columns]